src/main/kotlin/pt/up/fe/ni/website/backend/service/RoleService.kt

```diff
@@ -1,46 +1,113 @@
-        val foundActivity = activity.associatedRoles
+    fun grantPermissionToRoleOnActivity(roleId: Long, activityId: Long, permissions: Permissions) {
+        val activity = activityService.getActivityById(activityId)
+        val role = getRole(roleId)
+        val foundPerActivityRole = activity.associatedRoles
             .find { it.activity == activity } ?: PerActivityRole(Permissions())
-        foundActivity.role = role
-        foundActivity.activity = activity
+        foundPerActivityRole.role = role
+        foundPerActivityRole.activity = activity
 
-        foundActivity.role = role
-        foundActivity.permissions.add(permission)
-        perActivityRoleRepository.save(foundActivity)
+        foundPerActivityRole.permissions.addAll(permissions)
+        perActivityRoleRepository.save(foundPerActivityRole)
+        if (activity.associatedRoles.find { it.activity == activity } == null) {
+            role.associatedActivities.add(foundPerActivityRole)
+        }
+        // activityRepository.save(activity)
+        roleRepository.save(role)
     }
 
-    fun revokePermissionFromRoleOnActivity(role: Role, activity: Activity, permission: Permission) {
+    fun revokePermissionFromRoleOnActivity(roleId: Long, activityId: Long, permissions: Permissions) {
+        val activity = activityService.getActivityById(activityId)
+        val role = getRole(roleId)
         val foundActivity = activity.associatedRoles
             .find { it.role == role } ?: return
-
-        foundActivity.permissions.remove(permission)
+        foundActivity.permissions.removeAll(permissions)
         perActivityRoleRepository.save(foundActivity)
+        // activityRepository.save(activity)
+    }
+
+    fun createNewRole(dto: RoleDto): Role {
+        if (roleRepository.findByName(dto.name) != null) {
+            throw IllegalArgumentException(ErrorMessages.roleAlreadyExists)
+        }
+        val role = dto.create()
+
+        val latestGeneration = generationRepository.findFirstByOrderBySchoolYearDesc()
+            ?: throw IllegalArgumentException(ErrorMessages.noGenerations)
+
+        // need to set it from both sides due to testing transaction
+        role.generation = latestGeneration
+        roleRepository.save(role)
+        latestGeneration.roles.add(role)
+        generationRepository.save(latestGeneration)
+        return role
+    }
+
+    fun deleteRole(roleId: Long) {
+        val role = getRole(roleId)
+        role.generation.roles.remove(role)
+        generationRepository.save(role.generation)
+        roleRepository.delete(role)
+    }
+
+    fun addUserToRole(roleId: Long, userId: Long) {
+        val role = getRole(roleId)
+        val account = accountService.getAccountById(userId)
+        role.accounts.find { it.id == account.id }.let {
+            if (it != null) throw NoSuchElementException(ErrorMessages.userAlreadyHasRole(roleId, userId))
+        }
```

Comment:
In my opinion, this operation and the `removeUserFromRole` should be idempotent, thus returning 200 if the user already has the role. Do you find any helpful use case for this error?

Reply:
I thought of it being useful for displaying an error in the frontend if the user is already in role (maybe I should change the exception to make it a more useful HTTP code), it can also be idempotent but this verification must be then be done on the client side which adds a bit of unnecessary complexity, wdyt?

Reply:
I see 2 cases in the frontend:
- The user already has a role so the UI would be displaying the right info and the request wouldn't happen
- The UI is out of sync (e.g. bad network, someone else updated it) so the FE requests to add a role, receives 200 and updates the UI accordingly. Nothing really changes in the server but the result is as excepted for the API user.

Reply:
I agree, I'll change them to make them idempotent